Graphics/URP LOD Cross Fade

alekseiunity opened the pull request:
---
### Purpose of this PR
Currently in URP Mesh LOD switch inside LODGroup occurs immediately which frustrates the users as this ‘immediate’ switch looks like a sharp pop of a new object.
To improve that we need a mechanism to smooth transition or crossfade from one mesh to another. This could be achieved by using alpha test dithering mask crossfade https://en.wikipedia.org/wiki/Dither. Dithering mask, by using clip() or discard() function in a Pixel Shader, decreases total number of visible pixels of one Mesh LOD and increases total number of visible pixels of another Mesh LOD and smoothly crossfades two Meshes.

I propose to make 3 types of LOD Cross Fade dithering and allow user to choose the type in UniversalRenderPipelineAsset.

<img src="https://user-images.githubusercontent.com/43284133/149582846-1be0c1cb-a5ff-4c5a-94b5-81bd7cf38730.png" width="300">

LOD Cross Fade types are:

1. Bayer Matrix Dither
This is used in Built In renderer and very cheap but has quite repetitive pattern.

<img src="https://user-images.githubusercontent.com/43284133/149581927-2fb030de-f01e-44ee-9aaf-0be8c9a2e4da.png" width="700" height="700">

2. Blue Noise Dither
This uses precomputed blue noise texture and provides best looking option but a bit more expensive than Bayer Matrix.

<img src="https://user-images.githubusercontent.com/43284133/149581938-7f0f5764-8706-4237-9cd4-c1bb4c65825f.png" width="700" height="700">

3. Hashed Dither
This is used in HDRP and looks like a white noise with quite high discrepancy but main benefit of this option is that
dithering can be made camera rotation independent.

<img src="https://user-images.githubusercontent.com/43284133/149581946-9b0c400d-d068-47be-a510-9ef4ba9e6e14.png" width="700" height="700">

Test Scene Video: No Cross Fade - https://drive.google.com/file/d/1McO6CH33eNrBj3sCBtoOQHEUdXAuTdgf/view?usp=sharing
Test Scene Video: With Cross Fade - https://drive.google.com/file/d/1GNEjY3xh3DgzCslKawmfWGcwG3IDIHdp/view?usp=sharing


---
### Testing status
Tested on PC, Mac, Android, iOS and PS5 using my own synthetic scene with lots of objects.
Performance results are here. https://docs.google.com/spreadsheets/d/1r4EmXtUQMD3EYhFsDjfA2AE1tcx-MzzC110StR_guII/edit?usp=sharing

---
### Comments to reviewers

---
#### Shader permutations problem
Due to the limit on number of shader permutations we can't afford to select Cross Fade Type based on a keyword like:

#pragma multi_compile _ LOD_FADE_CROSSFADE_TYPE_BAYER_MATRIX

(Or can we?)

For this reason we select Cross Fade type based on const buffer uniform switch 'int _LODCrossFadeType'.
This also creates some problems for instance:

Hashed Dither can and should be made camera rotation independent. In order to do that we have to take viewVector in world space convert it in octahedral float2 and use that as a seed for Hash function. 

This is how it is done in HDRP:
````
uint2 ComputeFadeMaskSeed(float3 V, uint2 positionSS)
{
    uint2 fadeMaskSeed;
    if (IsPerspectiveProjection())
    {
        // Start with the world-space direction V. It is independent from the orientation of the camera,
        // and only depends on the position of the camera and the position of the fragment.
        // Now, project and transform it into [-1, 1].
        float2 pv = PackNormalOctQuadEncode(V);
        // Rescale it to account for the resolution of the screen.
        pv *= _ScreenParams.xy;
        // The camera only sees a small portion of the sphere, limited by hFoV and vFoV.
        // Therefore, we must rescale again (before quantization), roughly, by 1/tan(FoV/2).
        pv *= UNITY_MATRIX_P._m00_m11;
        // Truncate and quantize.
        fadeMaskSeed = asuint((int2)pv);
    }
    else
    {
        // Can't use the view direction, it is the same across the entire screen.
        fadeMaskSeed = positionSS;
    }

    return fadeMaskSeed;
}

LODDitheringTransition(ComputeFadeMaskSeed(positionSS), unity_LODFade.x);
````
So I have to prepare viewVector in every shader. And I can't compile out that code using a keyword because we use uniform as a switch.
Moreover we could end up with larger worst case VGPRs usage even if the user uses cheaper BayerMatrix type.

I don't know how to solve that. 

---
#### Additional Alpha To Coverage improvement
When MSAA is enabled in forward renderer. We can additionally use Alpha To Coverage to improve transition between LOD even more by writing dithering value to alpha. In that case transition looks a bit like alpha blended. 
We can integrate that later as soon as https://github.com/Unity-Technologies/Graphics/pull/6312 will be merged.

alekseiunity commented on com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs:
```diff
@@ -1234,6 +1255,18 @@ static void UpgradeAsset(int assetInstanceID)
             }
 
+            if(asset.k_AssetPreviousVersion < 11)
+            {
```
Is this a correct way to Reload PostProcessingData asset that now has to have links to blue noise and bayer matrix texture?

foxmalderalex replied:
> Is this a correct way to Reload PostProcessingData asset that now has to have links to blue noise and bayer matrix texture?

Moved textures to the asset itself as post processing data might not be assigned at all.

alekseiunity commented on com.unity.render-pipelines.universal/Shaders/BakedLitDepthNormalsPass.hlsl:
```diff
@@ -59,6 +58,8 @@ float4 DepthNormalsFragment(Varyings input) : SV_TARGET
     half alpha = texColor.a * _BaseColor.a;
     AlphaDiscard(alpha, _Cutoff);
 
+    LODFadeCrossFade(input.positionCS);
```
For some reason I have to put this always after alpha discard otherwise there is a weird behavior on iOS.

Aydindeveloper commented:
Hi. this is a great feature!
I wonder if you can make additions to use Stochastic Transparency with TAA.
Something like Crossfade Node. with alpha value. A node independent of the LOD. It can be added in Vibration for TAA. I used this to make a Deferred Clear coat with TAA by looking at the URP post processing Dithering shader. TAA + Stochastic Transparency would be a nice feature for URP.

![image](https://user-images.githubusercontent.com/49202311/151185577-5c4e41db-61b0-49a4-87b6-d56e8a802a45.png)

Verasl commented on com.unity.render-pipelines.universal/Runtime/Data/PostProcessData.asset.meta:
```diff
@@ -1,5 +1,5 @@
 fileFormatVersion: 2
-guid: 41439944d30ece34e96484bdb6645b55
+guid: ee973005a97ef0b4e8e4207d9609e05c
```
This really needs to be reverted :P it'll mess up everyones projects otherwise

alekseiunity replied:
Good point

sebastienlagarde commented:
Just a comment mainly for Universal team.

>Due to the limit on number of shader permutations we can't afford to select Cross Fade Type based on a keyword like:
>#pragma multi_compile _ LOD_FADE_CROSSFADE_TYPE_BAYER_MATRIX
>(Or can we?)

in HDRP LOD_FADE is a build time killer as it affect all our passes (both deferred adn forward, shadow, depth prepass etc...). So it is the worse thing that we can do for build time and sadly with current fading algorithm use in unity we have no choice.
So for sure you don't want to add more variant with extra option, so good you used constant buffer. But you should also really consider the addition of an option to strip those variant in the RP settings.

in HDRP we have added a supportDitheringCrossFade to allow to discard all those extra variant AND on ShaderGraph we have added an option to generate or not this multicompile (sadly we can't do this with non shader graph shader).

So for reviewer of this PR, be really consicous about the build time involve by this change! :)

alekseiunity commented:
> Just a comment mainly for Universal team.
> 
> > Due to the limit on number of shader permutations we can't afford to select Cross Fade Type based on a keyword like:
> > #pragma multi_compile _ LOD_FADE_CROSSFADE_TYPE_BAYER_MATRIX
> > (Or can we?)
> 
> in HDRP LOD_FADE is a build time killer as it affect all our passes (both deferred adn forward, shadow, depth prepass etc...). So it is the worse thing that we can do for build time and sadly with current fading algorithm use in unity we have no choice. So for sure you don't want to add more variant with extra option, so good you used constant buffer. But you should also really consider the addition of an option to strip those variant in the RP settings.
> 
> in HDRP we have added a supportDitheringCrossFade to allow to discard all those extra variant AND on ShaderGraph we have added an option to generate or not this multicompile (sadly we can't do this with non shader graph shader).
> 
> So for reviewer of this PR, be really consicous about the build time involve by this change! :)

Yeah I think adding shader variant strip makes sense. Will add it.

simon-engelbrecht-soerensen reviewed (approved):
Looking good PR doc: https://docs.google.com/document/d/1nVoioD996jynbWwN4ZCXjNPquMvkd6wZrkdUouVR6DY
